lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,362 @@
+// Copyright 2023 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:collection/collection.dart';
+
+import '../importer.dart';
+import './utils.dart';
+import 'dart:convert';
+import '../io.dart';
+import 'package:path/path.dart' as p;
+
+/// A filesystem importer to use for load implementation details, and for
+/// canonicalizing paths not defined in package.json.
+final _filesystemImporter = FilesystemImporter.cwd;
+
+/// An [Importer] that resolves `pkg:` URLs using the Node resolution algorithm.
+class NodePackageImporterInternal extends Importer {
+  final Uri entryPointURL;
+
+  /// Creates a Node Package Importer with the associated entry point url
+  NodePackageImporterInternal(this.entryPointURL);
+
+  @override
+  bool isNonCanonicalScheme(String scheme) {
+    return scheme == 'pkg';
+  }
+
+  @override
+  Uri? canonicalize(Uri url) {
+    if (url.scheme == 'file') return _filesystemImporter.canonicalize(url);
+    if (url.scheme != 'pkg') return null;
+
+    if (url.hasAuthority) {
+      throw "pkg: URL $url must not have a host, port, username or password.";
+    }
+    if (p.isAbsolute(url.path)) {
+      throw "pkg: URL $url must not be an absolute path.";
+    }
+    if (url.path.isEmpty) {
+      throw "pkg: URL $url must not have an empty path.";
+    }
+    if (url.hasQuery || url.hasFragment) {
+      throw "pkg: URL $url must not have a query or fragment.";
+    }
+
+    var baseURL = containingUrl?.scheme == 'file'
+        ? Uri.parse(containingUrl!.toFilePath())
```

Comment:
My conclusion after testing:

- The illegal character was a `:`, because the `containingUrl` being passed to `Uri.directory()` was e.g. `file:///D:/a/dart-sass/dart-sass/sass-spec/spec/sandbox/11582968143023309/js-api-spec.ts`
- It seems like `Uri.directory` is expecting a path, not a full URL with the `file:///` scheme. One hack that got the test passing in Windows was to replace `p.dirname(p.fromUri(Uri.directory(baseURL.toString())))` with `p.dirname(p.fromUri(Uri.directory(baseURL.path.substring(1))))`.
- A better solution seems to be to skip the call to `Uri.directory()` entirely, even in Windows -- `p.fromUri(baseURL)` seems to work fine in Windows as well.

See [full diff here](https://github.com/sass/dart-sass/pull/2130/files/ed585906ec0ee6b57518d83d03c5e872e6958402..b106be19bbe98cb89561c00f171fad7ac60cae28).